### ISSUE TITLE
Expose parameters to client for controlling retry diff delay

### ISSUE THREAD

**Issue by ShrikantJadhav-Ntnx** [nutanix/papiea#643]
After moving to papiea version 0.9.27, BMaaS started seeing slowness in system tests. With earlier papiea version (0.9.6) system tests used to get finished in 5-6 mins, and with papiea version 0.9.27 the same set of system tests started taking 20-25 mins. 

- From logs we noticed that most of the time was being spent in the state transiting; 
- Jaeger traces shown that state_diff_handlers are getting executed after every ~10+ seconds. 

So, we thought of reducing the retry diff delay using below config parameters; 
- We tried setting PAPIEA_DEBUG=true, and kind.diff_retry_exponent to values between 0 to 1 
- We tried setting PAPIEA_DIFF_RESOLVE_DELAY and PAPIEA_INTENT_RESOLVE_DELAY to 0 to 1 
however reducing these values did not had much effect on the time taken by system tests. tests were still taking around 20-25 mins.

Then as an experiment, just to root cause the issue - we checked out the papiea code and hardcoded diff_delay values to 0.1 at following parts of the diff_resolver.ts

```
intentful_engine/diff_resolver.ts 
@@ -85,7 +85,8 @@ export class DiffResolver {
         } else {
             return {
                 delay: {
-                    delay_seconds: this.calculateBackoffFn(retries, DiffResolver.MAXIMUM_BACKOFF, this.entropyFn(kind.diff_delay), kind.diff_retry_exponent, kind.name),
+                    //delay_seconds: this.calculateBackoffFn(retries, DiffResolver.MAXIMUM_BACKOFF, this.entropyFn(kind.diff_delay), kind.diff_retry_exponent, kind.name),
+                    delay_seconds: 0.1,
                     delay_set_time: new Date()
                 },
                 retries
@@ -94,7 +95,7 @@ export class DiffResolver {
     }
     private incrementDiffBackoff(backoff: Backoff, delay: Delay | null | undefined, kind: Kind): Backoff {
-        const retries = backoff.retries + 1
+        const retries = backoff.retries + 0.1
         return this.createDiffBackoff(kind, delay, retries)
     }
@@ -135,7 +136,8 @@ export class DiffResolver {
         if (result.data !== null && result.data !== undefined && result.data.delay_secs !== undefined
             && result.data.delay_secs !== null && !Number.isNaN(result.data.delay_secs)) {
             return {
-                delay_seconds: result.data.delay_secs,
+                //delay_seconds: result.data.delay_secs,
+                delay_seconds: 0.1,
                 delay_set_time: new Date()
             }
```

With this change, system tests run time was again down to ~6-7 mins. 

Would like to request the team to check if these parameters could be exposed to the client so that client could control this diff delay. 
So for system test environment client could set these parameters to very low values to increase the development speed and for production environment it could be default values set by the papiea internally.

Note: Corresponding BMaaS Jira: # https://jira.nutanix.com/browse/CLSTR-1463

**Comment by shlomi-nx** [comment]
Not sure what exactly you mean by expose to the client, but surely this needs to be easily configurable using env vars set up in docker-compose.yaml. Will this be sufficient?

**Comment by shlomi-nx** [comment]
@ShrikantJadhav-Ntnx When making the above changes, do you know which of the three locations is responsible for the improvement?

**Comment by ShrikantJadhav-Ntnx** [comment]
> Not sure what exactly you mean by expose to the client, but surely this needs to be easily configurable using env vars set up in docker-compose.yaml. Will this be sufficient?

By expose i meant, was either to allow setting it per kind like 'diff_retry_exponent' or having config parameter which could be set in docker-compose.

Yes, @shlomi-nx configuring by settiing env vars in docker-compose.yaml would be sufficient.  Thank you.

**Comment by ShrikantJadhav-Ntnx** [comment]
> @ShrikantJadhav-Ntnx When making the above changes, do you know which of the three locations is responsible for the improvement?

Yes, Almost all of the improvement were seen after doing only below change

```
@@ -135,7 +136,8 @@ export class DiffResolver {
         if (result.data !== null && result.data !== undefined && result.data.delay_secs !== undefined
             && result.data.delay_secs !== null && !Number.isNaN(result.data.delay_secs)) {
             return {
-                delay_seconds: result.data.delay_secs,
+                //delay_seconds: result.data.delay_secs,
+                delay_seconds: 0.1,
                 delay_set_time: new Date()
             }
```

**Comment by shlomi-nx** [comment]
> Yes, Almost all of the improvement were seen after doing only below change
> 
> ```
> @@ -135,7 +136,8 @@ export class DiffResolver {
>          if (result.data !== null && result.data !== undefined && result.data.delay_secs !== undefined
>              && result.data.delay_secs !== null && !Number.isNaN(result.data.delay_secs)) {
>              return {
> -                delay_seconds: result.data.delay_secs,
> +                //delay_seconds: result.data.delay_secs,
> +                delay_seconds: 0.1,
>                  delay_set_time: new Date()
>              }
> ```

This means that what helped was overriding the return value of the invoked handler, which should be configured by the provider and if not a default is used. We should be able to allow an env-var knob for default delay if provider returned nothing, but we must not override actual values returned by the provider. If the provider wishes to make these configurable, such as with a multiplier, it will need to do it specifically.

cc: @Ntnx-Igor-Korotach

**Comment by Emulebest** [comment]
> > Yes, Almost all of the improvement were seen after doing only below change
> > ```
> > @@ -135,7 +136,8 @@ export class DiffResolver {
> >          if (result.data !== null && result.data !== undefined && result.data.delay_secs !== undefined
> >              && result.data.delay_secs !== null && !Number.isNaN(result.data.delay_secs)) {
> >              return {
> > -                delay_seconds: result.data.delay_secs,
> > +                //delay_seconds: result.data.delay_secs,
> > +                delay_seconds: 0.1,
> >                  delay_set_time: new Date()
> >              }
> > ```
> 
> This means that what helped was overriding the return value of the invoked handler, which should be configured by the provider and if not a default is used. We should be able to allow an env-var knob for default delay if provider returned nothing, but we must not override actual values returned by the provider. If the provider wishes to make these configurable, such as with a multiplier, it will need to do it specifically.
> 
> cc: @Ntnx-Igor-Korotach

We changed the default knob to be responsible for the exponential multiplier. Maybe add a setting to turn multiplication off completely?
